### PR TITLE
hugo/feature/Create CoreInterruptIn

### PR DESCRIPTION
**Pull request by HPezz**
- :building_construction: (Interface): Add InterruptIn interface
- :sparkles: (CoreInterruptIn): Create leka CoreInterruptIn wrapper
- :white_check_mark: (CoreInterruptIn): Add CoreInterruptIn_test

### CODE REVIEW

File: drivers/CoreInterruptIn/source/CoreInterruptIn.cpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+{
+	_on_rise_callback = callback;
+	_interrupt_in.rise(mbed::Callback<void()> {[this] { _on_rise_callback(); }});
+}
+
+void CoreInterruptIn::fall(std::function<void()> const &callback)
+{
+	_on_fall_callback = callback;
+	_interrupt_in.fall(mbed::Callback<void()> {[this] { _on_fall_callback(); }});
+}
```
Comment (YannLocatelli) [inline drivers/CoreInterruptIn/source/CoreInterruptIn.cpp:24]: Regarding recent discovery, maybe should call the callback in event queue

```suggestion
void CoreInterruptIn::rise(std::function<void()> const &callback)
{
	_on_rise_callback = callback;
	_interrupt_in.rise(mbed::Callback<void()> {[this] { _event_queue.call( [this] { _on_rise_callback(); } ) }});
}

void CoreInterruptIn::fall(std::function<void()> const &callback)
{
	_on_fall_callback = callback;
	_interrupt_in.fall(mbed::Callback<void()> {[this] { _event_queue.call( [this] { _on_fall_callback(); } ) }});
}
```

File: drivers/CoreInterruptIn/include/CoreInterruptIn.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+#pragma once
+
+#include "interface/drivers/InterruptIn.h"
+
+namespace leka {
+
+class CoreInterruptIn : public interface::InterruptIn
+{
+  public:
+	explicit CoreInterruptIn(PinName pin) : _pin(pin) {};
```
Comment (YannLocatelli) [inline drivers/CoreInterruptIn/include/CoreInterruptIn.h:14]: Does this work?

```suggestion
	explicit CoreInterruptIn(PinName pin) : _interrupt_in(pin) {};
```

File: drivers/CoreInterruptIn/include/CoreInterruptIn.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	explicit CoreInterruptIn(PinName pin) : _interrupt_in(pin) { _event_queue.dispatch_forever(); };
+
+	auto read() -> int final;
+
+	void rise(std::function<void()> const &callback) final;
+	void fall(std::function<void()> const &callback) final;
+
+  private:
+	CoreEventQueue _event_queue;
+	mbed::InterruptIn _interrupt_in;
```
Comment (ladislas) [inline drivers/CoreInterruptIn/include/CoreInterruptIn.h:24]: as there is only one pin:

```suggestion
	explicit CoreInterruptIn(PinName pin) : _pin(pin) { _event_queue.dispatch_forever(); };

	auto read() -> int final;

	void rise(std::function<void()> const &callback) final;
	void fall(std::function<void()> const &callback) final;

  private:
	CoreEventQueue _event_queue;
	mbed::InterruptIn _pin;
```

File: drivers/CoreInterruptIn/include/CoreInterruptIn.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  public:
+	explicit CoreInterruptIn(PinName pin) : _interrupt_in(pin) { _event_queue.dispatch_forever(); };
+
+	auto read() -> int final;
+
+	void rise(std::function<void()> const &callback) final;
+	void fall(std::function<void()> const &callback) final;
+
+  private:
+	CoreEventQueue _event_queue;
```
Comment (ladislas) [inline drivers/CoreInterruptIn/include/CoreInterruptIn.h:23]: as explained on slack, this should be removed.

interrupts are very low level hardware primitive used to react **instantly** when they trigger.

event queues are used to differ execution of slow parts of code outside of ISR context.

the fact that we need to use an event queue after an interrupt is an implementation detail and should be handled by the developper.

CoreInterruptIn must be kept as low level and fast as possible.

File: include/interface/drivers/InterruptIn.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+class InterruptIn
+{
+  public:
+	virtual ~InterruptIn() = default;
+
+	virtual auto read() -> int = 0;
+
+	virtual void rise(std::function<void()> const &callback) = 0;
+	virtual void fall(std::function<void()> const &callback) = 0;
```
Comment (ladislas) [inline include/interface/drivers/InterruptIn.h:24]: i would add an `on`

```suggestion
	virtual void onRise(std::function<void()> const &callback) = 0;
	virtual void onFall(std::function<void()> const &callback) = 0;
```

**Review by ladislas — CHANGES_REQUESTED** [review]
LGTM 👍 
except for the CoreEventQueue which must be removed

**Review by ladislas — APPROVED** [review]
LGTM 👍

**Pull request merged**